### PR TITLE
Fix empty attribute projections for TagHelpers.

### PR DESCRIPTION
**Pull request by NTaylorMullen**
- Added TagHelperParseTreeRewriter tests, attempted to cover all empty attribute edge cases.
- Added Codegen tests to validate output and DesignTimeLineMappings.

#271

### CODE REVIEW

File: src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs [hunk 1]
```diff
@@ -134,11 +137,14 @@ public static class TagHelperBlockRewriter
                     SourceLocation symbolStartLocation;
 
                     // Check for attribute start values, aka single or double quote
-                    if (IsQuote(htmlSymbols[i + 1]))
+                    if (i + 1 < htmlSymbols.Length && IsQuote(htmlSymbols[i + 1]))
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs]: nit: parentheses around first `i + 1` to avoid readers having to think

File: src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs [hunk 2]
```diff
@@ -148,13 +154,17 @@ public static class TagHelperBlockRewriter
                         symbolOffset = 0;
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs]: why is this necessary?

File: src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs [hunk 3]
```diff
@@ -148,13 +154,17 @@ public static class TagHelperBlockRewriter
...
+                        span.Start +
+                        symbolStartLocation +
+                        new SourceLocation(absoluteIndex: 1, lineIndex: 0, characterIndex: 1);
+
                     afterEquals = true;
                 }
+                else if (symbol.Type == HtmlSymbolType.WhiteSpace)
+                {
+                    attributeValueStartLocation = SourceLocation.Advance(attributeValueStartLocation, symbol.Content);
+                }
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs]: what is this block for?  expect whitespace before and after the equals sign to be silently skipped.

File: src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs [hunk 4]
```diff
@@ -119,7 +122,7 @@ public static class TagHelperBlockRewriter
                 else if (name == null && symbol.Type == HtmlSymbolType.Text)
                 {
                     name = symbol.Content;
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs]: this method is complex enough to need comments at the top of almost every block explaining the current (or at least expected) situation.  e.g. `// Found the start of the name,` here

File: src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs [hunk 5]
```diff
@@ -119,7 +122,7 @@ public static class TagHelperBlockRewriter
...
-                    attributeValueStartLocation = SourceLocation.Advance(span.Start, name);
+                    attributeValueStartLocation = SourceLocation.Advance(attributeValueStartLocation, name);
                 }
                 else if (symbol.Type == HtmlSymbolType.Equals)
                 {
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs]: probably a long-standing bug but this will get very confused if the user typed `<input type==`.  should add `!afterEquals` to the condition.

File: src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs (hunk 1 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs]: this condition isn't in the right spot because whitespace is allowed between the equals sign and the opening quote or value.  IOW this method needs more states in its machine or to make existing states easier to see:
1. before name (skip whitespace if preceding code hasn't already done that)
2. name found (now skip whitespace)
3. equals found (now skip whitespace)
4. bare or quoted value found (perhaps need to skip whitespace after end of value or the matching quote

side question which branches in this method will `checked`, `@multiple`, `type = @type`, and `type = "@type"` hit?  realize `@multiple` isn't currently supported but I'm curious nonetheless.

File: test/Microsoft.AspNet.Razor.Test/TagHelpers/TagHelperParseTreeRewriterTest.cs [hunk 6]
```diff
@@ -19,6 +19,77 @@ namespace Microsoft.AspNet.Razor.Test.TagHelpers
...
+                            new MarkupTagHelperBlock("p",
+                            new Dictionary<string, SyntaxTreeNode>
+                            {
+                                { "class1",  new MarkupBlock() },
+                                { "class2",  factory.Markup("").With(SpanCodeGenerator.Null) },
+                                { "class3",  new MarkupBlock() },
+                            }))
+                    },
+                    {
+                        "<p class1=''class2=\"\"class3= />",
```
Comment (dougbu) [inline test/Microsoft.AspNet.Razor.Test/TagHelpers/TagHelperParseTreeRewriterTest.cs]: parser doesn't catch the error here?  space between attribute name/value pairs is required, not optional.

File: test/Microsoft.AspNet.Razor.Test/TestFiles/CodeGenerator/CS/Source/EmptyAttributeTagHelpers.cshtml [hunk 7]
```diff
@@ -0,0 +1,8 @@
+﻿@addtaghelper "something"
+
+<div>
+    <input type= checked=""class="" />
```
Comment (dougbu) [inline test/Microsoft.AspNet.Razor.Test/TestFiles/CodeGenerator/CS/Source/EmptyAttributeTagHelpers.cshtml]: not positive parser should accept mistake (missing whitespace) here or on line 6

File: test/Microsoft.AspNet.Razor.Test/Generator/CSharpTagHelperRenderingTest.cs [hunk 8]
```diff
@@ -213,7 +213,41 @@ public static TheoryData DesignTimeTagHelperTestData
                             BuildLineMapping(1094, 30, 18, 5179, 187, 19, 29),
                             BuildLineMapping(1231, 34, 5279, 192, 0, 1),
                         }
-                    }
+                    },
```
Comment (dougbu) [inline test/Microsoft.AspNet.Razor.Test/Generator/CSharpTagHelperRenderingTest.cs]: don't see tests for anything but bound `string` attributes.  what about unbound attributes and non-`string` bound attributes?  or am I missing something?

**Comment by dougbu** [comment]
:watch:

File: src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs (hunk 1 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs]: Right now Razor doesn't understand attributes with space around the equals as mentioned on lines 131/132 in this method.

1. `checked` will hit the second branch where we find the name. We will also do work to understand that we didn't find a value.
1. `@multiple` wont be in this method since that will be wrapped in a Block and will be handled by the `TryParseBlock` method.
1. `type = @type` will hit the other method `TryParseBlock` because it's understood as having a "dynamic" value.
1. type = "@type" same as above

File: src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs (hunk 2 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs]: We don't want to accept the end `"` or `'` into the newly created attribute value.

File: src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs (hunk 3 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs]: This block handles the leading whitespace of attributes. For example the block we're parsing is contained within the | characters: `<input| asp-for="..."| />`. Therefore this branch handles the very first characters of whitespace prior to the attribute name.

File: src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs (hunk 2 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs]: true but `symbolOffset` is initialized to `0`. so why set it to `0` again?

File: src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs (hunk 5 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs]: `afterEquals` is the first condition in the tree. It'll never get here :smile:

File: test/Microsoft.AspNet.Razor.Test/TagHelpers/TagHelperParseTreeRewriterTest.cs (hunk 6 above)
Comment (NTaylorMullen) [inline test/Microsoft.AspNet.Razor.Test/TagHelpers/TagHelperParseTreeRewriterTest.cs]: lmao, nope, the Razor parser for some odd reason decided to understand this format but not the mainstream format of having optional spaces around the equal sign.

File: test/Microsoft.AspNet.Razor.Test/TestFiles/CodeGenerator/CS/Source/EmptyAttributeTagHelpers.cshtml (hunk 7 above)
Comment (NTaylorMullen) [inline test/Microsoft.AspNet.Razor.Test/TestFiles/CodeGenerator/CS/Source/EmptyAttributeTagHelpers.cshtml]: Outstanding issue of the Razor parser not fully understanding attributes :sob:

File: test/Microsoft.AspNet.Razor.Test/Generator/CSharpTagHelperRenderingTest.cs (hunk 8 above)
Comment (NTaylorMullen) [inline test/Microsoft.AspNet.Razor.Test/Generator/CSharpTagHelperRenderingTest.cs]: `<p>` tags age attribute is an `int`

File: src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs (hunk 4 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs]: :+1: will add comments at the top of each branch stating where the cursor currently points and why.

File: src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs (hunk 3 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs]: well, it'll also handle whitespace between the attribute name and the equals sign.

more generally, what does eating the whitespace prior to the name have to do with empty (quoted in valid cases) attribute values?  (and, why did anything work without this block?)

File: src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs (hunk 2 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs]: Ahh good catch, I reversed the setting this time around to be set to 0 instead of 1 initially. :+1:

File: src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs (hunk 3 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs]: You definitely have a point but if there is space between the attribute name and the equals currently it wont even be understood as an attribute by the Razor parser and will then never make it here :disappointed: . That issue is supposed to be addressed by #123

**Comment by NTaylorMullen** [comment]
Addressed code review comments.

File: src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs [hunk 9]
```diff
@@ -104,6 +107,11 @@ public static class TagHelperBlockRewriter
 
                 if (afterEquals)
                 {
+                    // We've captured all leading whitespace, the attribute name, and an equals with an optional 
+                    // quote/double quote. We're now at: " asp-for='|...'" or " asp-for=|..."
+                    // The goal here is to capture all symbols until the end of the attribute. Note this will not 
+                    // consume an ending quote due to the symbolOffset.
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs]: teensy nit: a second note that whitespace between the attribute value or the closing quote and the next attribute or the end of the tag is not included in `htmlSymbols` may be worthwhile.

File: src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs [hunk 10]
```diff
@@ -118,11 +126,19 @@ public static class TagHelperBlockRewriter
...
+                    // We're now at: " |asp-for='...'" or " |asp-for=..."
+                    // The goal here is to capture the attribute name.
+
                     name = symbol.Content;
-                    attributeValueStartLocation = SourceLocation.Advance(span.Start, name);
+                    attributeValueStartLocation = SourceLocation.Advance(attributeValueStartLocation, name);
                 }
                 else if (symbol.Type == HtmlSymbolType.Equals)
                 {
+                    // We've captured all leading whitespace and the attribute name.
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs]: if you're sure the parser guarantees a name before the equals, add a `Debug.Assert(name != null)` to the top of this block

File: src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs [hunk 11]
```diff
@@ -118,11 +126,19 @@ public static class TagHelperBlockRewriter
...
+                    // We're now at: " asp-for|='...'" or " asp-for|=..."
+                    // The goal here is to consume the equal sign and the optional single/double-quote.
+
                     // We've found an '=' symbol, this means that the coming symbols will either be a quote
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs]: previous paragraph now describes the situation better.  Start with "The coming symbols ..."

File: src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs [hunk 12]
```diff
@@ -134,27 +150,37 @@ public static class TagHelperBlockRewriter
                     SourceLocation symbolStartLocation;
 
                     // Check for attribute start values, aka single or double quote
-                    if (IsQuote(htmlSymbols[i + 1]))
+                    if ((i + 1) < htmlSymbols.Length && IsQuote(htmlSymbols[i + 1]))
                     {
                         // Move past the attribute start so we can accept the true value.
                         i++;
-                        symbolStartLocation = htmlSymbols[i + 1].Start;
+                        symbolStartLocation = htmlSymbols[i].Start;
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs]: nit: `htmlSymbols[i]` -> `symbol`

File: src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs [hunk 13]
```diff
@@ -134,27 +150,37 @@ public static class TagHelperBlockRewriter
...
+                        symbolStartLocation +
+                        new SourceLocation(absoluteIndex: 1, lineIndex: 0, characterIndex: 1);
+
                     afterEquals = true;
                 }
+                else if (symbol.Type == HtmlSymbolType.WhiteSpace)
+                {
+                    // We're at the start of the attribute, this branch will always be hit on the first iterations of 
+                    // the loop. We're at: "| asp-for='...'" or "| asp-for=..."
+                    // Note: This will not be hit even for situations like asp-for  ="..." because the core Razor
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs]: nit: ~~even~~

File: src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs [hunk 14]
```diff
@@ -134,27 +150,37 @@ public static class TagHelperBlockRewriter
...
+                    attributeValueStartLocation = 
+                        span.Start +
+                        symbolStartLocation +
+                        new SourceLocation(absoluteIndex: 1, lineIndex: 0, characterIndex: 1);
+
                     afterEquals = true;
                 }
+                else if (symbol.Type == HtmlSymbolType.WhiteSpace)
+                {
+                    // We're at the start of the attribute, this branch will always be hit on the first iterations of
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs]: nit: ~~always~~ since the parser (strangely) separates adjacent attribute name/value pairs

File: test/Microsoft.AspNet.Razor.Test/TestFiles/CodeGenerator/CS/Output/EmptyAttributeTagHelpers.DesignTime.cs [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+            __InputTagHelper2 = CreateTagHelper<InputTagHelper2>();
+            __InputTagHelper2.Type = __InputTagHelper.Type;
+#line 6 "EmptyAttributeTagHelpers.cshtml"
+  __InputTagHelper2.Checked = ;
+
+#line default
+#line hidden
+            __PTagHelper = CreateTagHelper<PTagHelper>();
+#line 5 "EmptyAttributeTagHelpers.cshtml"
+__PTagHelper.Age = ;
```
Comment (dougbu) [inline test/Microsoft.AspNet.Razor.Test/TestFiles/CodeGenerator/CS/Output/EmptyAttributeTagHelpers.DesignTime.cs]: side note: do we have a bug about catching this error prior to compilation?  pretty easy for the c# writers to detect this case.

File: test/Microsoft.AspNet.Razor.Test/TestFiles/CodeGenerator/CS/Output/EmptyAttributeTagHelpers.DesignTime.cs [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+
+        #pragma warning disable 1998
+        public override async Task ExecuteAsync()
+        {
+            __InputTagHelper = CreateTagHelper<InputTagHelper>();
+            __InputTagHelper.Type = "";
+            __InputTagHelper2 = CreateTagHelper<InputTagHelper2>();
+            __InputTagHelper2.Type = __InputTagHelper.Type;
+#line 4 "EmptyAttributeTagHelpers.cshtml"
+__InputTagHelper2.Checked = ;
```
Comment (dougbu) [inline test/Microsoft.AspNet.Razor.Test/TestFiles/CodeGenerator/CS/Output/EmptyAttributeTagHelpers.DesignTime.cs]: this case is valid HTML.  should handle convert empty `bool` values to `true`.  (`checked`, `checked=''`, and `checked='checked'` all mean `true`; everything else is `false`) .

otherwise, we should warn tag helper authors that Razor isn't smart enough to handle HTML setting `bool` tag helper properties correctly.  suggest they bind them as `string`s and handle the meaning themselves.  that's what the MVC tag helpers currently do (and at least one get's it wrong).

File: test/Microsoft.AspNet.Razor.Test/TestFiles/CodeGenerator/CS/Source/EmptyAttributeTagHelpers.cshtml [hunk 17]
```diff
@@ -0,0 +1,8 @@
...
+    <p age=''>
+        <input type=""checked= class="" />
```
Comment (dougbu) [inline test/Microsoft.AspNet.Razor.Test/TestFiles/CodeGenerator/CS/Source/EmptyAttributeTagHelpers.cshtml]: this line contains an error (no value for `checked`).  why does code get generated?

File: test/Microsoft.AspNet.Razor.Test/TagHelpers/TagHelperParseTreeRewriterTest.cs (hunk 6 above)
Comment (dougbu) [inline test/Microsoft.AspNet.Razor.Test/TagHelpers/TagHelperParseTreeRewriterTest.cs]: could you open a bug about this lameness please?  or reference existing bug (if any) near the test case.

File: test/Microsoft.AspNet.Razor.Test/Generator/CSharpTagHelperRenderingTest.cs (hunk 8 above)
Comment (dougbu) [inline test/Microsoft.AspNet.Razor.Test/Generator/CSharpTagHelperRenderingTest.cs]: :ok:

File: src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs (hunk 5 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs]: right, sorry; it's just whitespace after equals, missing name, and missing value cases that could trip this code up.  and at least some of those (whitespace) won't reach here due to parser issues.  know missing value _does_ lead to code gen.  what about missing name?

**Comment by dougbu** [comment]
:watch:

File: test/Microsoft.AspNet.Razor.Test/TagHelpers/TagHelperParseTreeRewriterTest.cs (hunk 6 above)
Comment (NTaylorMullen) [inline test/Microsoft.AspNet.Razor.Test/TagHelpers/TagHelperParseTreeRewriterTest.cs]: https://github.com/aspnet/Razor/issues/287

File: src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs (hunk 9 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs]: The method itself handles a single attribute and therefore could never contain the whitespace. I don't think adding extra comment details here would add value. It might actually be confusing given the context of the method.

File: src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs (hunk 12 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs]: :+1:

File: test/Microsoft.AspNet.Razor.Test/TestFiles/CodeGenerator/CS/Output/EmptyAttributeTagHelpers.DesignTime.cs (hunk 15 above)
Comment (NTaylorMullen) [inline test/Microsoft.AspNet.Razor.Test/TestFiles/CodeGenerator/CS/Output/EmptyAttributeTagHelpers.DesignTime.cs]: No we don't I figured the compilation error would be good enough.

File: test/Microsoft.AspNet.Razor.Test/TestFiles/CodeGenerator/CS/Output/EmptyAttributeTagHelpers.DesignTime.cs (hunk 16 above)
Comment (NTaylorMullen) [inline test/Microsoft.AspNet.Razor.Test/TestFiles/CodeGenerator/CS/Output/EmptyAttributeTagHelpers.DesignTime.cs]: The attribute is bound so I'm not sure if the same mechanics apply, even if it is valid in HTML. In the case that you want your `TagHelper` to have dual functionality (augmenting an existing HTML tag and choosing to or not to use it as a `TagHelper`) then what you're stating definitely makes sense. 

As for the MVC `TagHelper`s we already prefix all `TagHelper` bound attributes to avoid ambiguity so thankfully we don't really have the problem (in my opinion). I don't think we should enable a case such as what you're suggesting because we'd then lose IntelliSense/accurate error handling. Definitely discussion worthy though.

**Note: ** Nothing actionable here on this comment (unless I'm missing something).

File: test/Microsoft.AspNet.Razor.Test/TestFiles/CodeGenerator/CS/Source/EmptyAttributeTagHelpers.cshtml (hunk 17 above)
Comment (NTaylorMullen) [inline test/Microsoft.AspNet.Razor.Test/TestFiles/CodeGenerator/CS/Source/EmptyAttributeTagHelpers.cshtml]: It's not an error according to the Razor parser. it's essentially parsing the attribute as if it was going to be represented like: `<input checked=true />` minus the value.

File: test/Microsoft.AspNet.Razor.Test/TestFiles/CodeGenerator/CS/Output/EmptyAttributeTagHelpers.DesignTime.cs (hunk 15 above)
Comment (dougbu) [inline test/Microsoft.AspNet.Razor.Test/TestFiles/CodeGenerator/CS/Output/EmptyAttributeTagHelpers.DesignTime.cs]: should not be letting errors we _hope_ the compiler will catch slip through.  have more information on the actual issue in Razor code.  for example, could let user know about specific invalid empty attribute where compiler can just say the generated code is bogus.  particularly important because compiler errors generally highlight the original Razor source and don't show the generated code.

File: src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs (hunk 14 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs]: Done on line underneath this.

File: test/Microsoft.AspNet.Razor.Test/TestFiles/CodeGenerator/CS/Source/EmptyAttributeTagHelpers.cshtml (hunk 17 above)
Comment (dougbu) [inline test/Microsoft.AspNet.Razor.Test/TestFiles/CodeGenerator/CS/Source/EmptyAttributeTagHelpers.cshtml]: it's another invalid case that leads to bogus C#.  might work when `checked` is not bound or bound to a `string` property.  but broken in the context of tag helpers.
not immediately actionable but we need to have a design discussion about errors in tag helper bindings.  probably worth an issue to track that discussion.

File: test/Microsoft.AspNet.Razor.Test/TestFiles/CodeGenerator/CS/Output/EmptyAttributeTagHelpers.DesignTime.cs (hunk 15 above)
Comment (NTaylorMullen) [inline test/Microsoft.AspNet.Razor.Test/TestFiles/CodeGenerator/CS/Output/EmptyAttributeTagHelpers.DesignTime.cs]: https://github.com/aspnet/Razor/issues/289

File: test/Microsoft.AspNet.Razor.Test/TestFiles/CodeGenerator/CS/Source/EmptyAttributeTagHelpers.cshtml (hunk 17 above)
Comment (NTaylorMullen) [inline test/Microsoft.AspNet.Razor.Test/TestFiles/CodeGenerator/CS/Source/EmptyAttributeTagHelpers.cshtml]: The issue I just created would apply: https://github.com/aspnet/Razor/issues/289

File: test/Microsoft.AspNet.Razor.Test/TestFiles/CodeGenerator/CS/Output/EmptyAttributeTagHelpers.DesignTime.cs (hunk 16 above)
Comment (dougbu) [inline test/Microsoft.AspNet.Razor.Test/TestFiles/CodeGenerator/CS/Output/EmptyAttributeTagHelpers.DesignTime.cs]: agreed, not immediately actionable.  but we should have an issue and design discussion about `bool` tag helper properties and how we should handle them.  I suspect the main question is what would be most useful to tag helper authors?  (FYI when writing `<select/>` tag helper, I went out of my way to define `Multiple` as `string` since Razor wouldn't help if it was a `bool`.)

File: test/Microsoft.AspNet.Razor.Test/TestFiles/CodeGenerator/CS/Source/EmptyAttributeTagHelpers.cshtml (hunk 17 above)
Comment (dougbu) [inline test/Microsoft.AspNet.Razor.Test/TestFiles/CodeGenerator/CS/Source/EmptyAttributeTagHelpers.cshtml]: :+1:

File: src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs (hunk 14 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs]: huh?  my point was "always" isn't true

File: src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs (hunk 9 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs]: agree this method handles a single attribute.  however the content of `htmlSymbols` isn't obvious.  might just need a comment near where it's defined that mentions it may contain preceding whitespace, will always contain the attribute name, will likely contain equals and a value, and will not contain appended whitespace.  or a shorter comment subsetting that info :smile:

File: src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs (hunk 14 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs]: no, my point was "always" is not true

File: src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs [hunk 18]
```diff
@@ -134,27 +154,38 @@ public static class TagHelperBlockRewriter
                     SourceLocation symbolStartLocation;
 
                     // Check for attribute start values, aka single or double quote
-                    if (IsQuote(htmlSymbols[i + 1]))
+                    if ((i + 1) < htmlSymbols.Length && IsQuote(htmlSymbols[i + 1]))
                     {
                         // Move past the attribute start so we can accept the true value.
                         i++;
-                        symbolStartLocation = htmlSymbols[i + 1].Start;
+                        symbolStartLocation = symbol.Start;
```
Comment (dougbu) [inline src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs]: sorry, I was wrong here.  `symbol != htmlSymbols[i]` anymore.  you were right to spell `htmlSymbols[i].Start` out.  (if I'd been correct, would have worked to assign `symbolStartLocation` above the `if`.)

**Comment by dougbu** [comment]
:shipit: with one incorrect suggestion I made reversed and (maybe) a code comment or two improved

File: src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs (hunk 14 above)
Comment (NTaylorMullen) [inline src/Microsoft.AspNet.Razor/Parser/TagHelpers/TagHelperBlockRewriter.cs]: :+1:
